### PR TITLE
Print the index and summary of the SampleInput that failed an OpInfo test

### PR DESCRIPTION
**Pull request by jbschlosser**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #99444

Related to the Reproducible Testing BE project. Goal is to print out the sample input that failed an OpInfo test.

Crazy idea: to avoid requiring widespread changes across tests that use OpInfo sample inputs, return a new special iterator type from `OpInfo.sample_inputs()`, etc. that tracks the most recent item seen. If a test fails later on, print out this info to identify the sample that failed the test.

This solves the problem that the test framework currently has no concept of which sample input is being operated on.

This PR contains the following changes:
* New `TrackedInputIter` that wraps a sample inputs func iterator and tracks the most recent input seen in a `TrackedInput` structure
    * The information is stored in a dictionary on the test function itself, mapping `full test ID -> most recent TrackedInput`
* To determine the test function that is being run, we do some stack crawling hackery in `extract_test_fn_and_id()`
* Above applies only when one of the following is called: `OpInfo.sample_inputs()`, `OpInfo.error_inputs()`, `OpInfo.reference_inputs()`, and `OpInfo.conjugate_sample_inputs()`. This could easily be extended to `ModuleInfo`s and the sparse sample input funcs as well

Example output when a sample input causes a failure:
```
======================================================================
ERROR: test_foo_add_cpu_uint8 (__main__.TestFakeTensorCPU)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jbschlosser/branches/reproducible_testing/torch/testing/_internal/common_device_type.py", line 911, in test_wrapper
    return test(*args, **kwargs)
  File "/home/jbschlosser/branches/reproducible_testing/torch/testing/_internal/common_device_type.py", line 1097, in only_fn
    return fn(slf, *args, **kwargs)
  File "/home/jbschlosser/branches/reproducible_testing/test/test_ops.py", line 2211, in test_foo
    self.fail('Example failure')
AssertionError: Example failure

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/jbschlosser/branches/reproducible_testing/torch/testing/_internal/common_utils.py", line 2436, in wrapper
    method(*args, **kwargs)
  File "/home/jbschlosser/branches/reproducible_testing/torch/testing/_internal/common_device_type.py", line 414, in instantiated_test
    result = test(self, **param_kwargs)
  File "/home/jbschlosser/branches/reproducible_testing/torch/testing/_internal/common_device_type.py", line 917, in test_wrapper
    raise Exception(
Exception: Caused by sample input at index 2: SampleInput(input=Tensor[size=(5, 1), device="cpu", dtype=torch.uint8], args=TensorList[Tensor[size=(5,), device="cpu", dtype=torch.uint8]], kwargs={}, broadcasts_input=True, name='')

To execute this test, run the following from the base repo dir:
     python test/test_ops.py -k test_foo_add_cpu_uint8

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0

----------------------------------------------------------------------
```

This notably doesn't print the actual `SampleInput` values, as that's hard without fully reproducible random sample generation. I went down this path for a while and it seems infeasible without adding an untenable amount of overhead to set the random seed per SampleInput (see https://github.com/pytorch/pytorch/issues/86694#issuecomment-1614943708 for more details). For now, I am settling for at least spitting out the index and some metadata of the `SampleInput`, as it seems better than nothing.

### CODE REVIEW

File: torch/testing/_internal/common_device_type.py [hunk 1]
```diff
@@ -843,7 +843,20 @@ def _parametrize_test(self, test, generic_cls, device_cls):
...
+                            if tracked_input is not None:
+                                raise Exception(
+                                    f"Error caused by {tracked_input.input_type} "
+                                    f"{tracked_input.index}: {tracked_input.input_val}") from e
+                            raise e
+
+                    # Initialize info for the last inputs seen. This is useful for tracking
+                    # down which inputs caused a test failure.
+                    # Mapping format: test ID -> TrackedInput
+                    test.tracked_inputs = dict()
```
Comment (albanD) [inline torch/testing/_internal/common_device_type.py:859]: Maybe mention here that the iterator is responsible for populating and cleaning this field to make sure you don't use extra memory.

File: torch/testing/_internal/common_device_type.py [hunk 2]
```diff
@@ -843,7 +843,20 @@ def _parametrize_test(self, test, generic_cls, device_cls):
...
+                            tracked_input = get_tracked_input()
+                            if tracked_input is not None:
+                                raise Exception(
+                                    f"Error caused by {tracked_input.input_type} "
+                                    f"{tracked_input.index}: {tracked_input.input_val}") from e
+                            raise e
+
+                    # Initialize info for the last inputs seen. This is useful for tracking
+                    # down which inputs caused a test failure.
+                    # Mapping format: test ID -> TrackedInput
```
Comment (albanD) [inline torch/testing/_internal/common_device_type.py:858]: Why is this needed? Aren't you only ever running one test at a time anyways?

**Review by albanD — COMMENTED** [review]
Pretty cool!

File: torch/testing/_internal/common_device_type.py (hunk 2 above)
Comment (jbschlosser) [reply torch/testing/_internal/common_device_type.py:858]: This was written under the assumption that we could be running tests across multiple threads, but I'm not sure if that's true?

File: torch/testing/_internal/common_device_type.py [hunk 3]
```diff
@@ -843,7 +843,20 @@ def _parametrize_test(self, test, generic_cls, device_cls):
...
-                        return test(*args, **kwargs)
+                        try:
+                            return test(*args, **kwargs)
+                        except Exception as e:
+                            tracked_input = get_tracked_input()
+                            if tracked_input is not None:
+                                raise Exception(
+                                    f"Error caused by {tracked_input.input_type} "
+                                    f"{tracked_input.index}: {tracked_input.input_val}") from e
+                            raise e
```
Comment (zou3519) [inline torch/testing/_internal/common_device_type.py:854]: This is pretty funny, and I am OK with the stack-crawling approach we took here.

A feature request is: there should be some way to, given this TrackedInput information,  actually be able to reproduce the test locally. One idea for how to do this, taken from Hypothesis, is to have some sort of decorator that we could apply to a function:
```py
# On a local build, we can apply the following "example" decorator.
# It modifies the test to only pull the specific OpInfo and the index of the sample input
@example(opinfo="torch.add", sample_input_index=3)
@ops(op_db)
def test_vmap(self, device, dtype, op):
    for sample_input in op.sample_inputs():
        run_test(sample_input)
```

Comment (jbschlosser) [reply torch/testing/_internal/common_device_type.py:854]: Awesome, I like that `@example` syntax! I think something like that should be doable; lemme see what I can throw together.

File: torch/testing/_internal/common_utils.py [hunk 4]
```diff
@@ -135,6 +136,87 @@ def wrapper(*args, **kwargs):
         fn(*args, **kwargs)
     return wrapper
 
+# Tries to extract the test function and full test ID of the current function under test
+# by crawling the stack. If unsuccessful, returns a pair of Nones.
+def extract_test_fn_and_id() -> Optional[Tuple[Callable, str]]:
```
Comment (bdhirsh) [inline torch/testing/_internal/common_utils.py:141]: nit: should the return type be `Tuple[Optional[Callable], Optional[str]]`?

Comment (jbschlosser) [reply torch/testing/_internal/common_utils.py:141]: yep sure enough!

**Comment by janeyx99** [comment]
@jbschlosser how's BE day going

File: torch/testing/_internal/opinfo/core.py [hunk 5]
```diff
@@ -227,8 +227,15 @@ def formatter(arg):
...
-                return f"Tensor[{shape}]"
+                shape = str(tuple(arg.shape))
+                dtype = str(arg.dtype)
+                device = str(arg.device)
+                contiguity_suffix = ""
+                # NB: sparse CSR tensors annoyingly return is_sparse=False
+                is_sparse = arg.is_sparse or arg.layout == torch.sparse_csr
+                if not is_sparse and not arg.is_contiguous():
+                    contiguity_suffix = ", contiguous=False"
+                return f'Tensor[size={shape}, device="{device}", dtype={dtype}{contiguity_suffix}]'
```
Comment (jbschlosser) [inline torch/testing/_internal/opinfo/core.py:238]: This part makes the `SampleInput.summary()` output a bit nicer and more useful

**Review by janeyx99 — APPROVED** [review]
Seems useful! Thanks :)

I am curious if the TrackedInputIter could be simplified if we know that the tests are running in different processes (vs threads) based on @albanD's comment in April.

File: torch/testing/_internal/common_device_type.py (hunk 2 above)
Comment (clee2000) [reply torch/testing/_internal/common_device_type.py:931]: yes thats correct we do run tests in parallel across different processes

**Comment by huydhn** [comment]
@pytorchbot revert -m 'Sorry for reverting your change but it seems to cause memory leak on CUDA job https://hud.pytorch.org/pytorch/pytorch/commit/e7f12b1eb0cedfd20dcb41ea35e21e9a71e3390a' -c nosignal

Tests are run in parallel, so CUDA memory leak check isn't run on PR, but only periodically

**Comment by jbschlosser** [comment]
For posterity: I moved the responsibility for clearing out the tracked input from the `TrackedInputIter` -> the test wrapper in a finally clause. For tests that only run on a single sample (and thus do not exhaust the iterator), the old way was causing the memory leaks that got this PR reverted.
